Eventually should report what's going wrong together with the attempts

Given the following piece of code:

```
init {
        "Foo" {
            eventually(3.seconds, AssertionError::class.java) {
                3 shouldBe 3
                1 shouldBe 2
                4 shouldBe 4
            }
        }
    }
```

It's a code that will test a lot of assertions. In this specific case, I mocked some assertions with one of them being wrong, but escalate that to a real world scenario.

This test obviously fails, as `1 shouldBe 2` will throw an AssertionError. However, when we take a look at the failure message, it doesn't tell us what failed:

` java.lang.AssertionError: Test failed after 3 seconds; attempted 168628 times`

With this message, I'm not sure if the non-deterministic didn't complete (for example, an API didn't return), or the expected return didn't work.

----

To solve this issue, the only workaround is to debug the code line by line to try to find what happened for it to fail.

It would be better if `eventually` included the thrown exception message with it, including the stacktrace of that exception, so that the developer could pinpoint the error.

--- a/kotlintest-assertions/src/main/kotlin/io/kotlintest/Eventually.kt
+++ b/kotlintest-assertions/src/main/kotlin/io/kotlintest/Eventually.kt
@@ -8,6 +8,7 @@ fun <T> eventually(duration: Duration, f: () -> T): T = eventually(duration, Exc
 fun <T, E : Throwable> eventually(duration: Duration, exceptionClass: Class<E>, f: () -> T): T {
   val end = System.nanoTime() + duration.toNanos()
   var times = 0
+  var lastError: Throwable? = null
   while (System.nanoTime() < end) {
     try {
       return f()
@@ -16,11 +17,13 @@ fun <T, E : Throwable> eventually(duration: Duration, exceptionClass: Class<E>, 
         // Not the kind of exception we were prepared to tolerate
         throw e
       }
+      lastError = e
       // else ignore and continue
     }
     times++
   }
-  throw Failures.failure("Test failed after ${duration.seconds} seconds; attempted $times times")
+  val underlyingCause = if (lastError == null) "" else "; underlying cause was ${lastError.localizedMessage}"
+  throw Failures.failure("Test failed after ${duration.seconds} seconds; attempted $times times$underlyingCause")
 }
 
 fun <T> eventually(duration: Duration, predicate: (T) -> Boolean, f: () -> T): T {

--- a/kotlintest-assertions/src/main/kotlin/io/kotlintest/Eventually.kt
+++ b/kotlintest-assertions/src/main/kotlin/io/kotlintest/Eventually.kt
@@ -23,7 +23,7 @@ fun <T, E : Throwable> eventually(duration: Duration, exceptionClass: Class<E>, 
     times++
   }
   val underlyingCause = if (lastError == null) "" else "; underlying cause was ${lastError.localizedMessage}"
-  throw Failures.failure("Test failed after ${duration.seconds} seconds; attempted $times times$underlyingCause")
+  throw Failures.failure("Test failed after ${duration.seconds} seconds; attempted $times times$underlyingCause", lastError)
 }
 
 fun <T> eventually(duration: Duration, predicate: (T) -> Boolean, f: () -> T): T {

--- a/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/EventuallyTest.kt
+++ b/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/EventuallyTest.kt
@@ -74,7 +74,7 @@ class EventuallyTest : WordSpec() {
       }
       "display the underlying assertion failure" {
         shouldThrow<AssertionError> {
-          eventually(Duration.ofSeconds(2), AssertionError::class.java) {
+          eventually(Duration.ofMillis(10), AssertionError::class.java) {
             1 shouldBe 2
           }
         }.message.shouldEndWith("; underlying cause was expected: 2 but was: 1")

--- a/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/EventuallyTest.kt
+++ b/kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/EventuallyTest.kt
@@ -1,6 +1,7 @@
 package com.sksamuel.kotlintest
 
 import io.kotlintest.eventually
+import io.kotlintest.matchers.string.shouldEndWith
 import io.kotlintest.shouldBe
 import io.kotlintest.shouldThrow
 import io.kotlintest.specs.WordSpec
@@ -70,6 +71,13 @@ class EventuallyTest : WordSpec() {
           1
         }
         result shouldBe 1
+      }
+      "display the underlying assertion failure" {
+        shouldThrow<AssertionError> {
+          eventually(Duration.ofSeconds(2), AssertionError::class.java) {
+            1 shouldBe 2
+          }
+        }.message.shouldEndWith("; underlying cause was expected: 2 but was: 1")
       }
     }
   }

